internal/federationdomain/endpoints/callback/callback_handler.go

```diff
@@ -58,48 +57,31 @@ func NewHandler(
 		// an error if the client requested a scope that they are not allowed to request, so we don't need to worry about that here.
 		downstreamsession.AutoApproveScopes(authorizeRequester)
 
-		token, err := upstreamIDPConfig.ExchangeAuthcodeAndValidateTokens(
-			r.Context(),
-			authcode(r),
-			state.PKCECode,
-			state.Nonce,
-			redirectURI,
-		)
+		identity, loginExtras, err := idp.HandleCallback(r.Context(), authcode(r), state.PKCECode, state.Nonce, redirectURI)
 		if err != nil {
-			plog.WarningErr("error exchanging and validating upstream tokens", err, "upstreamName", upstreamIDPConfig.GetName())
-			return httperr.New(http.StatusBadGateway, "error exchanging and validating upstream tokens")
-		}
-
-		subject, upstreamUsername, upstreamGroups, err := downstreamsession.GetDownstreamIdentityFromUpstreamIDToken(
-			upstreamIDPConfig, token.IDToken.Claims, resolvedOIDCIdentityProvider.DisplayName,
-		)
-		if err != nil {
-			return httperr.Wrap(http.StatusUnprocessableEntity, err.Error(), err)
+			return err
 		}
 
 		username, groups, err := downstreamsession.ApplyIdentityTransformations(
-			r.Context(), resolvedOIDCIdentityProvider.Transforms, upstreamUsername, upstreamGroups,
-		)
-		if err != nil {
-			return httperr.Wrap(http.StatusUnprocessableEntity, err.Error(), err)
-		}
-
-		additionalClaims := downstreamsession.MapAdditionalClaimsFromUpstreamIDToken(upstreamIDPConfig, token.IDToken.Claims)
-
-		customSessionData, err := downstreamsession.MakeDownstreamOIDCCustomSessionData(
-			upstreamIDPConfig, token, username, upstreamUsername, upstreamGroups,
+			r.Context(), idp.GetTransforms(), identity.UpstreamUsername, identity.UpstreamGroups,
```

Comment:
While reveiwing together we noted that this `ApplyIdentityTransformations()` could be collapsed down into `NewPinnipedSession()` as these are currently (and always should be) used together.

Reply:
Done in the latest commit. This seems like a nice improvement.